nilearn/plotting/surf_plotting.py

```diff
@@ -29,10 +31,502 @@
+
+    # create face values from vertex values by selected avg methods
+    if avg_method == 'mean':
+        surf_map_faces = np.mean(surf_map_data[faces], axis=1)
+    elif avg_method == 'median':
+        surf_map_faces = np.median(surf_map_data[faces], axis=1)
+    elif avg_method == 'min':
+        surf_map_faces = np.min(surf_map_data[faces], axis=1)
+    elif avg_method == 'max':
+        surf_map_faces = np.max(surf_map_data[faces], axis=1)
+    elif callable(avg_method):
+        surf_map_faces = np.apply_along_axis(
+            avg_method, 1, surf_map_data[faces]
+        )
+
+        # check that surf_map_faces has the same length as face_colors
+        if surf_map_faces.shape != (face_colors_size,):
+            raise ValueError(
+                "Array computed with the custom function "
+                "from avg_method does not have the correct shape: "
+                f"{surf_map_faces[0]} != {face_colors_size}")
+
+        # check that dtype is either int or float
+        if not (
+            "int" in str(surf_map_faces.dtype)
+            or "float" in str(surf_map_faces.dtype)
+        ):
+            raise ValueError(
+                'Array computed with the custom function '
+                'from avg_method should be an array of numbers '
+                '(int or float)'
+            )
+    else:
+        raise ValueError(
+            "avg_method should be either "
+            "['mean', 'median', 'max', 'min'] "
+            "or a custom function"
+        )
+    return surf_map_faces
+
+
+def _configure_cmap_matplotlib(cmap, vmin, vmax, threshold,
+                               cbar_tick_format):
+    """Helper function for plot_surf with matplotlib engine.
+
+    This function configure the colormap.
+    """
+    our_cmap = get_cmap(cmap)
+    norm = Normalize(vmin=vmin, vmax=vmax)
+    # Default number of ticks is 5...
+    nb_ticks = 5
+    # ...unless we are dealing with integers with a small range
+    # in this case, we reduce the number of ticks
+    if cbar_tick_format == "%i" and vmax - vmin < nb_ticks:
+        ticks = np.arange(vmin, vmax + 1)
+        nb_ticks = len(ticks)
+    else:
+        ticks = np.linspace(vmin, vmax, nb_ticks)
+    bounds = np.linspace(vmin, vmax, our_cmap.N)
+    if threshold is not None:
+        cmaplist = [our_cmap(i) for i in range(our_cmap.N)]
+        # set colors to grey for absolute values < threshold
+        istart = int(norm(-threshold, clip=True) * (our_cmap.N - 1))
+        istop = int(norm(threshold, clip=True) * (our_cmap.N - 1))
+        for i in range(istart, istop):
+            cmaplist[i] = (0.5, 0.5, 0.5, 1.)
+        our_cmap = LinearSegmentedColormap.from_list(
+            'Custom cmap', cmaplist, our_cmap.N)
+    return our_cmap, norm, ticks, bounds
+
+
+def _compute_facecolors(bg_map, faces, n_vertices, darkness, alpha):
+    """Helper function for plot_surf.
+
+    This function computes the facecolors.
```

Comment:
Used for matplotlib rendering only, right ?

Reply:
Thanks @bthirion ! When I wrote it I wasn't sure if it would be matplotlib-specific or not.
As for this PR, it is only used when the engine is matplotlib since we are coloring by points rather than faces when using plotly. I will thus update the name and docstring.
That being said, I realized recently that it _could_ still be useful to have the possibility to color by faces with plotly. For example, when trying to extend to `plot_surf_contours`, it seems much easier and faster to draw the contours by coloring faces than by coloring points. I can explain in more details why I think that's the case (I might also be wrong here...).

Reply:
maybe having "matplotlib" in the name or at least the docstring would make it a bit easier to identify when this is used. most likely we will not want to color faces except when forced to do so when using matplotlib. but anyway we can always rename this function later if needed as it is "private"